File: 2_add_two_numbers/addTwoNumbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        root = n = ListNode(0)
        carry = 0

        while l1 or l2 or carry:
            value1 = value2 = 0
            if l1:
                value1 = l1.val
                l1 = l1.next
            if l2:
                value2 = l2.val
                l2 = l2.next
            summ = value1 + value2
            carry, value = divmod(summ + carry, 10)
            n.next = ListNode(value)
            n = n.next

        return root.next

File: 2_add_two_numbers/addTwoNumbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	root := new(ListNode)
	carry := 0

	for n := root; l1 != nil || l2 != nil || carry > 0; n = n.Next {
		if l1 != nil {
			carry += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			carry += l2.Val
			l2 = l2.Next
		}
		val := carry % 10
		n.Next = &ListNode{val, nil}
		carry /= 10
	}

	return root.Next
}